### PR TITLE
Separate controllers

### PR DESCRIPTION
**Pull request by gastonmaspero**
## Summary ##

**Trello card**: https://trello.com/c/nAAz4iNb/21-n-72-extraer-componentes-en-controllers-separados

Description: Small refactor. Book's comments and book's details have now their own separate controller, view, view model and xib.

### Screenshots ###

iPhone X

Simulator Screen Shot - iPhone X - 2019-06-07 at 13 29 33
Simulator Screen Shot - iPhone X - 2019-06-07 at 13 29 27

### CODE REVIEW

File: WBooks/Book Details/BookDetailsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    
+    @objc func rent() {
+        switch _viewModel.status {
+        case .available:
+            let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "THANKS".localized(), message: "PROCESSING_RENT".localized(), dismissButtonTitle: "ACCEPT".localized()))
+            self.present(alert, animated: true, completion: nil)
+            _viewModel.rentBook(onSuccess: onBookRentSuccess, onError: onBookRentError)
+        case .notAvailable, .inHands:
+            let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "UPS!", message: "CANNOT_RENT".localized(), dismissButtonTitle: "ACCEPT".localized()))
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewController.swift]: Add the string to Localizable

File: WBooks/Book Details/BookDetailsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            self.present(alert, animated: true, completion: nil)
+        case .notLoaded:
+            let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "UPS!", message: "RENT_ERROR".localized(), dismissButtonTitle: "ACCEPT".localized()))
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewController.swift]: Same

File: WBooks/Book Details/BookDetailsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    
+    func onBookRentSuccess() {
+        _viewModel.onBookRentSuccess()
+        let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "THANKS".localized(), message: "RENT_COMPLETED".localized(), dismissButtonTitle: "ACCEPT".localized()))
+        self.present(alert, animated: true, completion: nil)
+        _view.setAvailability(status: _viewModel.status)
+    }
+    
+    func onBookRentError() {
+        let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "Error!", message: "RENT_ERROR".localized(), dismissButtonTitle: "ACCEPT".localized()))
```
Comment (carcos-wolox) [inline WBooks/Book Details/BookDetailsViewController.swift]: Same

File: WBooks/Book Info/BookInfoViewController.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        //Book details
+        _bookDetailsController.view.translatesAutoresizingMaskIntoConstraints = false
+        _bookDetailsController.view.topAnchor.constraint(equalTo: topLayoutGuide.bottomAnchor, constant: 10).isActive = true
+        _bookDetailsController.view.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 20).isActive = true
+        _bookDetailsController.view.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -20).isActive = true
+        //Comments
+        _commentsController.view.topAnchor.constraint(equalTo: _bookDetailsController.view.bottomAnchor, constant: 10).isActive = true
+        _commentsController.view.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 20).isActive = true
+        _commentsController.view.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -20).isActive = true
+        _commentsController.view.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -20).isActive = true
```
Comment (carcos-wolox) [inline WBooks/Book Info/BookInfoViewController.swift]: `20` seems to be a value that you are using constantly. It is better to set this kind of values in a constant property, so if it changes in the time you only need to change ir once.

File: WBooks/Book Info/BookInfoViewController.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        view.addSubview(_commentsController.view)
+        setupLayout()
+    }
+    
+    private func setupLayout() {
+        //Book details
+        _bookDetailsController.view.translatesAutoresizingMaskIntoConstraints = false
+        _bookDetailsController.view.topAnchor.constraint(equalTo: topLayoutGuide.bottomAnchor, constant: 10).isActive = true
+        _bookDetailsController.view.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 20).isActive = true
+        _bookDetailsController.view.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -20).isActive = true
```
Comment (carcos-wolox) [inline WBooks/Book Info/BookInfoViewController.swift]: You could also activate this constraints doing: `NSLayoutConstraint.activate([_bookDetailsController.view.topAnchor.constraint(equalTo: topLayoutGuide.bottomAnchor, constant: 10), ...]), so you don't need to activate each of them

File: WBooks/Library/MainMenuController.swift [hunk 6]
```diff
@@ -58,7 +58,8 @@ extension MainMenuController: UITableViewDataSource, UITableViewDelegate {
     }
     
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
-        let controller = DetailViewController(viewModel: DetailViewModel(bookViewModel: _viewModel.books[indexPath.row]))
+        let bookViewModel = _viewModel.books[indexPath.row]
+        let controller = BookInfoViewController(viewModel: BookInfoViewModel(bookViewModel: bookViewModel), commentsController: BookCommentsViewController(viewModel: BookCommentsViewModel(bookViewModel: bookViewModel)), bookDetailsController: BookDetailsViewController(viewModel: BookDetailsViewModel(bookViewModel: bookViewModel)))
```
Comment (carcos-wolox) [inline WBooks/Library/MainMenuController.swift]: This line is too long, break it in different lines or create the viewControllers in different properties and use them here.

File: WBooks/Book Details/BookDetailsViewController.swift (hunk 1 above)
Comment (gastonmaspero) [reply WBooks/Book Details/BookDetailsViewController.swift]: You mean "UPS"? I decided to use that expression in all languages. Should I add it to Localizable anyway?

File: WBooks/Book Info/BookInfoView.xib [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (carcos-wolox) [inline WBooks/Book Info/BookInfoView.xib]: If the view does not have anything, I think you don't need the .xib nor the view's class, UIViewController has it's own view.

**Review by carcos-wolox — CHANGES_REQUESTED** [review]
In the screenshot the book's title is getting off the space; make it to take 2 lines or show ... if the string is too big. Also, comments view should take the rest of the Information view, you have a big empty space below comment's view

**Review by carcos-wolox — APPROVED** [review]
👍  Good!

**Pull request merged**